# Question 1: Word2Vec on Health & Fitness Sentences

In this notebook we train a Word2Vec model on 20 health and fitness sentences using **Gensim**,
and then look at the words the model thinks are similar.

## Step 0: Install and import libraries

In [1]:
!pip install gensim


[notice] A new release of pip is available: 24.2 -> 26.2.1
[notice] To update, run: pip install --upgrade pip


In [2]:
import string
import nltk
import pandas as pd
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from gensim.models import Word2Vec

nltk.download('punkt')
nltk.download('punkt_tab')
nltk.download('stopwords')

[nltk_data] Downloading package punkt to
[nltk_data]     /Users/sayanbhattacharya/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     /Users/sayanbhattacharya/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/sayanbhattacharya/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

## Step 1: Load the sentences into a Python list

In [3]:
sentences = [
    "A balanced diet helps maintain a healthy body weight",
    "Regular workout sessions improve overall muscle strength",
    "Drinking enough water daily supports good digestion",
    "A high protein diet is common among strength athletes",
    "Cardio workouts help improve heart and lung health",
    "Skipping meals can slow down your metabolism over time",
    "Stretching before a workout reduces the risk of injury",
    "A balanced diet should include fruits, vegetables and whole grains",
    "Strength training workouts build muscle and improve bone density",
    "Getting enough sleep is important for muscle recovery",
    "A sudden change in diet can cause digestive discomfort",
    "Yoga is a low-impact workout that improves flexibility",
    "Doctors recommend a balanced diet to prevent lifestyle diseases",
    "Interval training workouts burn calories quickly",
    "A high-sugar diet is linked to several health problems",
    "Consistent workout routines improve long-term fitness levels",
    "Nutritionists often design a diet plan based on individual goals",
    "Warm-up exercises prepare the body for an intense workout",
    "A vegetarian diet can still provide sufficient protein",
    "Rest days are an important part of any workout schedule"
]

print("Total number of sentences:", len(sentences))

Total number of sentences: 20


Save the corpus as a CSV file (needed for the submission).

In [4]:
df = pd.DataFrame({"text": sentences})
df.to_csv("word2vec_corpus.csv", index=False)
df.head()

,text
0,A balanced diet helps maintain a healthy body ...
1,Regular workout sessions improve overall muscl...
2,Drinking enough water daily supports good dige...
3,A high protein diet is common among strength a...
4,Cardio workouts help improve heart and lung he...


## Step 2: Convert to lowercase and remove punctuation

In [5]:
def clean_text(text):
    text = text.lower()
    for p in string.punctuation:
        text = text.replace(p, " ")   # replace punctuation with a space
    return text

cleaned_sentences = []
for s in sentences:
    cleaned_sentences.append(clean_text(s))

for s in cleaned_sentences[:5]:
    print(s)

a balanced diet helps maintain a healthy body weight
regular workout sessions improve overall muscle strength
drinking enough water daily supports good digestion
a high protein diet is common among strength athletes
cardio workouts help improve heart and lung health


## Step 3: Tokenize each sentence into words

In [6]:
tokenized_sentences = []
for s in cleaned_sentences:
    tokens = word_tokenize(s)
    tokenized_sentences.append(tokens)

for t in tokenized_sentences[:5]:
    print(t)

['a', 'balanced', 'diet', 'helps', 'maintain', 'a', 'healthy', 'body', 'weight']
['regular', 'workout', 'sessions', 'improve', 'overall', 'muscle', 'strength']
['drinking', 'enough', 'water', 'daily', 'supports', 'good', 'digestion']
['a', 'high', 'protein', 'diet', 'is', 'common', 'among', 'strength', 'athletes']
['cardio', 'workouts', 'help', 'improve', 'heart', 'and', 'lung', 'health']


## Step 4 (Optional): Remove stopwords

In [7]:
stop_words = set(stopwords.words('english'))

final_sentences = []
for tokens in tokenized_sentences:
    new_tokens = []
    for word in tokens:
        if word not in stop_words:
            new_tokens.append(word)
    final_sentences.append(new_tokens)

for t in final_sentences[:5]:
    print(t)

['balanced', 'diet', 'helps', 'maintain', 'healthy', 'body', 'weight']
['regular', 'workout', 'sessions', 'improve', 'overall', 'muscle', 'strength']
['drinking', 'enough', 'water', 'daily', 'supports', 'good', 'digestion']
['high', 'protein', 'diet', 'common', 'among', 'strength', 'athletes']
['cardio', 'workouts', 'help', 'improve', 'heart', 'lung', 'health']


In [8]:
# compare the number of unique words before and after stopword removal
words_before = set()
for tokens in tokenized_sentences:
    for w in tokens:
        words_before.add(w)

words_after = set()
for tokens in final_sentences:
    for w in tokens:
        words_after.add(w)

print("Unique words with stopwords   :", len(words_before))
print("Unique words without stopwords:", len(words_after))

Unique words with stopwords   : 122
Unique words without stopwords: 102


## Step 5: Train the Word2Vec model

The dataset is very small, so we use small values:
- `vector_size = 50` : each word becomes a list of 50 numbers
- `window = 3` : look at 3 words on the left and right
- `min_count = 1` : keep every word (even if it appears only once)
- `sg = 1` : use the skip-gram method (works better on small data)
- `epochs = 100` : go over the data many times because it is small
- `seed = 42` and `workers = 1` : so that we get the same result every time

In [9]:
model = Word2Vec(
    sentences=final_sentences,
    vector_size=50,
    window=3,
    min_count=1,
    sg=1,
    epochs=100,
    seed=42,
    workers=1
)

print("Model trained!")

Model trained!


## Step 6: Print the complete vocabulary

In [10]:
vocab = list(model.wv.index_to_key)

print("Total number of sentences:", len(sentences))
print("Vocabulary size:", len(vocab))
print()
print(vocab)

Total number of sentences: 20
Vocabulary size: 102

['diet', 'workout', 'improve', 'workouts', 'strength', 'muscle', 'balanced', 'important', 'training', 'health', 'protein', 'high', 'enough', 'body', 'schedule', 'part', 'days', 'rest', 'sufficient', 'provide', 'still', 'vegetarian', 'intense', 'prepare', 'exercises', 'warm', 'goals', 'individual', 'based', 'plan', 'design', 'often', 'nutritionists', 'levels', 'fitness', 'term', 'long', 'routines', 'consistent', 'problems', 'several', 'linked', 'sugar', 'quickly', 'calories', 'burn', 'interval', 'diseases', 'lifestyle', 'prevent', 'recommend', 'doctors', 'flexibility', 'improves', 'impact', 'low', 'yoga', 'discomfort', 'digestive', 'cause', 'change', 'sudden', 'recovery', 'sleep', 'getting', 'density', 'bone', 'build', 'grains', 'whole', 'vegetables', 'fruits', 'include', 'injury', 'risk', 'reduces', 'stretching', 'time', 'metabolism', 'slow', 'meals', 'skipping', 'lung', 'heart', 'help', 'cardio', 'athletes', 'among', 'common', 'diges

## Step 7: Word vector for "diet"

In [11]:
diet_vector = model.wv['diet']

print("Shape of the vector:", diet_vector.shape)
print()
print("First 10 values:")
print(diet_vector[:10])

Shape of the vector: (50,)

First 10 values:
[-0.02545768 -0.00867327  0.00423751 -0.00316094 -0.01257511  0.02292636
 -0.02652349  0.04279565 -0.03737486 -0.00703913]


In [12]:
print("Full vector for 'diet':")
print(diet_vector)

Full vector for 'diet':
[-0.02545768 -0.00867327  0.00423751 -0.00316094 -0.01257511  0.02292636
 -0.02652349  0.04279565 -0.03737486 -0.00703913 -0.01558221  0.04827917
 -0.00840396  0.03589156  0.03706356  0.01478708 -0.01585216 -0.04985502
  0.00947372 -0.00540012  0.00953812  0.00133883 -0.00629369  0.03325627
  0.04958679  0.00102725 -0.02828566 -0.00254329 -0.00377864 -0.0094717
 -0.00293793 -0.02196947 -0.0461235   0.01753233  0.00294092 -0.03195698
  0.01978445 -0.04920826 -0.00080506  0.03424937 -0.01213361  0.00041616
  0.04502997 -0.03477506 -0.00981525  0.01599505  0.01090025  0.03301707
  0.02598327 -0.00745531]


## Step 8: Top 5 words most similar to "diet"

In [13]:
similar_diet = model.wv.most_similar('diet', topn=5)

print("Top 5 words similar to 'diet':")
for word, score in similar_diet:
    print(word, ":", round(score, 4))

Top 5 words similar to 'diet':
improve : 0.7776
workout : 0.7495
based : 0.7402
enough : 0.7277
good : 0.7174


## Step 9: Top 5 words most similar to "workout"

In [14]:
similar_workout = model.wv.most_similar('workout', topn=5)

print("Top 5 words similar to 'workout':")
for word, score in similar_workout:
    print(word, ":", round(score, 4))

Top 5 words similar to 'workout':
enough : 0.835
improve : 0.8108
diet : 0.7495
workouts : 0.7464
muscle : 0.7144


Show both results in a small table using pandas.

In [15]:
table = pd.DataFrame({
    "diet - similar word": [w for w, s in similar_diet],
    "diet - score": [round(s, 4) for w, s in similar_diet],
    "workout - similar word": [w for w, s in similar_workout],
    "workout - score": [round(s, 4) for w, s in similar_workout]
})
table

,diet - similar word,diet - score,workout - similar word,workout - score
0,improve,0.7776,enough,0.8350
1,workout,0.7495,improve,0.8108
2,based,0.7402,diet,0.7495
3,enough,0.7277,workouts,0.7464
4,good,0.7174,muscle,0.7144


## Step 10: Interpretation

Some of the similar words make sense. For **workout**, the model found *workouts*, *muscle* and *improve*, which really
are related to exercise. For **diet**, words like *workout*, *improve* and *good* come up, which is partly sensible
(diet and workout are both about health), but words like *based* and *enough* are not really related to diet.
This happens because our dataset is very small (only 20 sentences and about 100 unique words), so most words appear
only once and the model does not see enough examples to learn good meanings. Also *diet* and *workout* appear in many
sentences, so they end up close to many common words. With a much bigger dataset the results would be cleaner.

## Questions for Interpretation

**1. Why is the `min_count` parameter needed, and what would happen if you increased it on this dataset?**

`min_count` tells the model to ignore words that appear fewer times than this number. In a big dataset this removes
rare words and spelling mistakes that the model cannot learn well anyway. In our small dataset most words appear
only once, so if we increase `min_count` the vocabulary becomes very small. As the check below shows, the vocabulary
drops from 102 words (min_count=1) to only 14 words (min_count=2) and 7 words (min_count=3), so we lose most of the words.

In [16]:
# small check: vocabulary size for different min_count values
for m in [1, 2, 3]:
    test_model = Word2Vec(sentences=final_sentences, vector_size=50, window=3,
                          min_count=m, sg=1, epochs=100, seed=42, workers=1)
    print("min_count =", m, "-> vocabulary size =", len(test_model.wv.index_to_key))

min_count = 1 -> vocabulary size = 102
min_count = 2 -> vocabulary size = 14
min_count = 3 -> vocabulary size = 7


**2. What does increasing the window size change about the relationships the model learns?**

The window is how many words on each side are treated as "context". A small window learns words that appear right
next to each other (more grammar-like / functional similarity). A larger window looks at more of the sentence, so the
model learns broader, topic-level relationships (for example, words from the same kind of sentence about diet or exercise).
In our short sentences a big window covers almost the whole sentence.

**3. Why might your similarity results differ slightly if you re-train the model?**

Word2Vec starts with random vectors and also uses random sampling (negative sampling) and, with several workers,
the order of training can change. So each training run can end up with slightly different vectors and scores.
With such a small dataset this effect is even bigger. Setting `seed` and `workers=1` helps to get the same result again.

# Question 2: Text Preprocessing Pipeline

Cleaning, Tokenization, Stopword Removal, Stemming (Porter & Snowball) and Lemmatization
on noisy health & fitness sentences using **pandas** and **NLTK**.

## Step 0: Import libraries

In [1]:
import re
import nltk
import pandas as pd
from nltk.tokenize import word_tokenize, RegexpTokenizer
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer, SnowballStemmer, WordNetLemmatizer

nltk.download('punkt')
nltk.download('punkt_tab')
nltk.download('stopwords')
nltk.download('wordnet')
nltk.download('omw-1.4')

[nltk_data] Downloading package punkt to
[nltk_data]     /Users/sayanbhattacharya/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     /Users/sayanbhattacharya/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/sayanbhattacharya/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     /Users/sayanbhattacharya/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     /Users/sayanbhattacharya/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


True

## Step 1: Load the raw sentences into a DataFrame

In [2]:
raw_sentences = [
    "A balanced DIET helps maintain healthy body weight!!",
    "regular workout sessions improve muscle strength... love it :)",
    "drinking 8 glasses of water daily supports digestion!!",
    "skipping meals can SLOW down your metabolism...",
    "stretching before a workout reduces injury risk!! important",
    "strength training builds muscle & improves bone density",
    "getting 7-8 hrs of sleep is important for recovery :)",
    "yoga is a LOW-impact workout, improves flexibility!!",
    "a high-sugar diet is linked to health problems...",
    "interval training burns calories QUICKLY!! great results",
    "nutritionists design diet plans based on individual goals",
    "rest days are an important part of any workout schedule!!"
]

df = pd.DataFrame({"text": raw_sentences})

# save the dataset as dataset.csv
df.to_csv("dataset.csv", index=False)

df

,text
0,A balanced DIET helps maintain healthy body we...
1,regular workout sessions improve muscle streng...
2,drinking 8 glasses of water daily supports dig...
3,skipping meals can SLOW down your metabolism...
4,stretching before a workout reduces injury ris...
5,strength training builds muscle & improves bon...
6,getting 7-8 hrs of sleep is important for reco...
7,"yoga is a LOW-impact workout, improves flexibi..."
8,a high-sugar diet is linked to health problems...
9,interval training burns calories QUICKLY!! gre...


## Step 2: Cleaning function

The function will:
1. make the text lowercase
2. remove punctuation, numbers and special characters (keep only letters and spaces)
3. remove extra spaces

In [3]:
def clean_text(text):
    text = text.lower()                       # 1. lowercase
    text = re.sub(r'[^a-z\s]', ' ', text)     # 2. keep only letters and spaces
    text = re.sub(r'\s+', ' ', text)          # 3. collapse extra spaces
    text = text.strip()
    return text

df["clean_text"] = df["text"].apply(clean_text)

Raw and cleaned text side by side:

In [4]:
pd.set_option('display.max_colwidth', None)
pd.set_option('display.max_rows', 100)
df[["text", "clean_text"]]

,text,clean_text
0,A balanced DIET helps maintain healthy body weight!!,a balanced diet helps maintain healthy body weight
1,regular workout sessions improve muscle strength... love it :),regular workout sessions improve muscle strength love it
2,drinking 8 glasses of water daily supports digestion!!,drinking glasses of water daily supports digestion
3,skipping meals can SLOW down your metabolism...,skipping meals can slow down your metabolism
4,stretching before a workout reduces injury risk!! important,stretching before a workout reduces injury risk important
5,strength training builds muscle & improves bone density,strength training builds muscle improves bone density
6,getting 7-8 hrs of sleep is important for recovery :),getting hrs of sleep is important for recovery
7,"yoga is a LOW-impact workout, improves flexibility!!",yoga is a low impact workout improves flexibility
8,a high-sugar diet is linked to health problems...,a high sugar diet is linked to health problems
9,interval training burns calories QUICKLY!! great results,interval training burns calories quickly great results


## Step 3: Tokenization (word_tokenize vs RegexpTokenizer)

In [5]:
# (a) NLTK word_tokenize
df["word_tokens"] = df["clean_text"].apply(word_tokenize)

# (b) RegexpTokenizer that keeps only alphabetic words
regexp_tokenizer = RegexpTokenizer(r'[a-zA-Z]+')
df["regexp_tokens"] = df["clean_text"].apply(regexp_tokenizer.tokenize)

for i in range(3):
    print("Sentence", i + 1, ":", df["clean_text"][i])
    print("word_tokenize   :", df["word_tokens"][i])
    print("RegexpTokenizer :", df["regexp_tokens"][i])
    print()

Sentence 1 : a balanced diet helps maintain healthy body weight
word_tokenize   : ['a', 'balanced', 'diet', 'helps', 'maintain', 'healthy', 'body', 'weight']
RegexpTokenizer : ['a', 'balanced', 'diet', 'helps', 'maintain', 'healthy', 'body', 'weight']

Sentence 2 : regular workout sessions improve muscle strength love it
word_tokenize   : ['regular', 'workout', 'sessions', 'improve', 'muscle', 'strength', 'love', 'it']
RegexpTokenizer : ['regular', 'workout', 'sessions', 'improve', 'muscle', 'strength', 'love', 'it']

Sentence 3 : drinking glasses of water daily supports digestion
word_tokenize   : ['drinking', 'glasses', 'of', 'water', 'daily', 'supports', 'digestion']
RegexpTokenizer : ['drinking', 'glasses', 'of', 'water', 'daily', 'supports', 'digestion']



Because we already cleaned the text (no punctuation or numbers left), both tokenizers give the **same** output here.
To see the real difference, let us also try both tokenizers on the **raw** text of the first 3 sentences:

In [6]:
for i in range(3):
    print("Raw sentence", i + 1, ":", df["text"][i])
    print("word_tokenize   :", word_tokenize(df["text"][i]))
    print("RegexpTokenizer :", regexp_tokenizer.tokenize(df["text"][i]))
    print()

Raw sentence 1 : A balanced DIET helps maintain healthy body weight!!
word_tokenize   : ['A', 'balanced', 'DIET', 'helps', 'maintain', 'healthy', 'body', 'weight', '!', '!']
RegexpTokenizer : ['A', 'balanced', 'DIET', 'helps', 'maintain', 'healthy', 'body', 'weight']

Raw sentence 2 : regular workout sessions improve muscle strength... love it :)
word_tokenize   : ['regular', 'workout', 'sessions', 'improve', 'muscle', 'strength', '...', 'love', 'it', ':', ')']
RegexpTokenizer : ['regular', 'workout', 'sessions', 'improve', 'muscle', 'strength', 'love', 'it']

Raw sentence 3 : drinking 8 glasses of water daily supports digestion!!
word_tokenize   : ['drinking', '8', 'glasses', 'of', 'water', 'daily', 'supports', 'digestion', '!', '!']
RegexpTokenizer : ['drinking', 'glasses', 'of', 'water', 'daily', 'supports', 'digestion']



**Difference:** On raw text, `word_tokenize` keeps punctuation (`!!`, `...`, `:`, `)`) and numbers (`8`) as separate
tokens, while `RegexpTokenizer` with the pattern `[a-zA-Z]+` keeps only the alphabetic words. On cleaned text both are the same.

## Step 4: Stopword removal

In [7]:
stop_words = set(stopwords.words('english'))

def remove_stopwords(tokens):
    result = []
    for word in tokens:
        if word not in stop_words:
            result.append(word)
    return result

df["no_stopwords"] = df["word_tokens"].apply(remove_stopwords)

df["count_before"] = df["word_tokens"].apply(len)
df["count_after"] = df["no_stopwords"].apply(len)

df[["clean_text", "count_before", "count_after"]]

,clean_text,count_before,count_after
0,a balanced diet helps maintain healthy body weight,8,7
1,regular workout sessions improve muscle strength love it,8,7
2,drinking glasses of water daily supports digestion,7,6
3,skipping meals can slow down your metabolism,7,4
4,stretching before a workout reduces injury risk important,8,6
5,strength training builds muscle improves bone density,7,7
6,getting hrs of sleep is important for recovery,8,5
7,yoga is a low impact workout improves flexibility,8,6
8,a high sugar diet is linked to health problems,9,6
9,interval training burns calories quickly great results,7,7


In [8]:
for i in range(len(df)):
    print(i + 1, df["no_stopwords"][i])

1 ['balanced', 'diet', 'helps', 'maintain', 'healthy', 'body', 'weight']
2 ['regular', 'workout', 'sessions', 'improve', 'muscle', 'strength', 'love']
3 ['drinking', 'glasses', 'water', 'daily', 'supports', 'digestion']
4 ['skipping', 'meals', 'slow', 'metabolism']
5 ['stretching', 'workout', 'reduces', 'injury', 'risk', 'important']
6 ['strength', 'training', 'builds', 'muscle', 'improves', 'bone', 'density']
7 ['getting', 'hrs', 'sleep', 'important', 'recovery']
8 ['yoga', 'low', 'impact', 'workout', 'improves', 'flexibility']
9 ['high', 'sugar', 'diet', 'linked', 'health', 'problems']
10 ['interval', 'training', 'burns', 'calories', 'quickly', 'great', 'results']
11 ['nutritionists', 'design', 'diet', 'plans', 'based', 'individual', 'goals']
12 ['rest', 'days', 'important', 'part', 'workout', 'schedule']


## Step 5: Porter Stemmer

In [9]:
porter = PorterStemmer()

def porter_stem(tokens):
    result = []
    for word in tokens:
        result.append(porter.stem(word))
    return result

df["porter"] = df["no_stopwords"].apply(porter_stem)

for i in range(len(df)):
    print(i + 1, df["porter"][i])

1 ['balanc', 'diet', 'help', 'maintain', 'healthi', 'bodi', 'weight']
2 ['regular', 'workout', 'session', 'improv', 'muscl', 'strength', 'love']
3 ['drink', 'glass', 'water', 'daili', 'support', 'digest']
4 ['skip', 'meal', 'slow', 'metabol']
5 ['stretch', 'workout', 'reduc', 'injuri', 'risk', 'import']
6 ['strength', 'train', 'build', 'muscl', 'improv', 'bone', 'densiti']
7 ['get', 'hr', 'sleep', 'import', 'recoveri']
8 ['yoga', 'low', 'impact', 'workout', 'improv', 'flexibl']
9 ['high', 'sugar', 'diet', 'link', 'health', 'problem']
10 ['interv', 'train', 'burn', 'calori', 'quickli', 'great', 'result']
11 ['nutritionist', 'design', 'diet', 'plan', 'base', 'individu', 'goal']
12 ['rest', 'day', 'import', 'part', 'workout', 'schedul']


## Step 6: Snowball Stemmer (English)

In [10]:
snowball = SnowballStemmer('english')

def snowball_stem(tokens):
    result = []
    for word in tokens:
        result.append(snowball.stem(word))
    return result

df["snowball"] = df["no_stopwords"].apply(snowball_stem)

for i in range(len(df)):
    print(i + 1, df["snowball"][i])

1 ['balanc', 'diet', 'help', 'maintain', 'healthi', 'bodi', 'weight']
2 ['regular', 'workout', 'session', 'improv', 'muscl', 'strength', 'love']
3 ['drink', 'glass', 'water', 'daili', 'support', 'digest']
4 ['skip', 'meal', 'slow', 'metabol']
5 ['stretch', 'workout', 'reduc', 'injuri', 'risk', 'import']
6 ['strength', 'train', 'build', 'muscl', 'improv', 'bone', 'densiti']
7 ['get', 'hrs', 'sleep', 'import', 'recoveri']
8 ['yoga', 'low', 'impact', 'workout', 'improv', 'flexibl']
9 ['high', 'sugar', 'diet', 'link', 'health', 'problem']
10 ['interv', 'train', 'burn', 'calori', 'quick', 'great', 'result']
11 ['nutritionist', 'design', 'diet', 'plan', 'base', 'individu', 'goal']
12 ['rest', 'day', 'import', 'part', 'workout', 'schedul']


## Step 7: WordNet Lemmatizer

By default the lemmatizer treats every word as a noun. Verbs like *improves* or *builds* would not change,
so we first try the word as a verb, and if nothing changes we try it as a noun.

In [11]:
lemmatizer = WordNetLemmatizer()

def lemmatize_word(word):
    lemma = lemmatizer.lemmatize(word, pos='v')    # try as a verb
    if lemma == word:
        lemma = lemmatizer.lemmatize(word, pos='n')  # try as a noun
    return lemma

def lemmatize_tokens(tokens):
    result = []
    for word in tokens:
        result.append(lemmatize_word(word))
    return result

df["lemma"] = df["no_stopwords"].apply(lemmatize_tokens)

for i in range(len(df)):
    print(i + 1, df["lemma"][i])

1 ['balance', 'diet', 'help', 'maintain', 'healthy', 'body', 'weight']
2 ['regular', 'workout', 'session', 'improve', 'muscle', 'strength', 'love']
3 ['drink', 'glass', 'water', 'daily', 'support', 'digestion']
4 ['skip', 'meal', 'slow', 'metabolism']
5 ['stretch', 'workout', 'reduce', 'injury', 'risk', 'important']
6 ['strength', 'train', 'build', 'muscle', 'improve', 'bone', 'density']
7 ['get', 'hr', 'sleep', 'important', 'recovery']
8 ['yoga', 'low', 'impact', 'workout', 'improve', 'flexibility']
9 ['high', 'sugar', 'diet', 'link', 'health', 'problem']
10 ['interval', 'train', 'burn', 'calorie', 'quickly', 'great', 'result']
11 ['nutritionist', 'design', 'diet', 'plan', 'base', 'individual', 'goal']
12 ['rest', 'day', 'important', 'part', 'workout', 'schedule']


All outputs side by side:

In [12]:
df[["no_stopwords", "porter", "snowball", "lemma"]]

,no_stopwords,porter,snowball,lemma
0,"[balanced, diet, helps, maintain, healthy, body, weight]","[balanc, diet, help, maintain, healthi, bodi, weight]","[balanc, diet, help, maintain, healthi, bodi, weight]","[balance, diet, help, maintain, healthy, body, weight]"
1,"[regular, workout, sessions, improve, muscle, strength, love]","[regular, workout, session, improv, muscl, strength, love]","[regular, workout, session, improv, muscl, strength, love]","[regular, workout, session, improve, muscle, strength, love]"
2,"[drinking, glasses, water, daily, supports, digestion]","[drink, glass, water, daili, support, digest]","[drink, glass, water, daili, support, digest]","[drink, glass, water, daily, support, digestion]"
3,"[skipping, meals, slow, metabolism]","[skip, meal, slow, metabol]","[skip, meal, slow, metabol]","[skip, meal, slow, metabolism]"
4,"[stretching, workout, reduces, injury, risk, important]","[stretch, workout, reduc, injuri, risk, import]","[stretch, workout, reduc, injuri, risk, import]","[stretch, workout, reduce, injury, risk, important]"
5,"[strength, training, builds, muscle, improves, bone, density]","[strength, train, build, muscl, improv, bone, densiti]","[strength, train, build, muscl, improv, bone, densiti]","[strength, train, build, muscle, improve, bone, density]"
6,"[getting, hrs, sleep, important, recovery]","[get, hr, sleep, import, recoveri]","[get, hrs, sleep, import, recoveri]","[get, hr, sleep, important, recovery]"
7,"[yoga, low, impact, workout, improves, flexibility]","[yoga, low, impact, workout, improv, flexibl]","[yoga, low, impact, workout, improv, flexibl]","[yoga, low, impact, workout, improve, flexibility]"
8,"[high, sugar, diet, linked, health, problems]","[high, sugar, diet, link, health, problem]","[high, sugar, diet, link, health, problem]","[high, sugar, diet, link, health, problem]"
9,"[interval, training, burns, calories, quickly, great, results]","[interv, train, burn, calori, quickli, great, result]","[interv, train, burn, calori, quick, great, result]","[interval, train, burn, calorie, quickly, great, result]"


## Step 8: Comparison table (Original | Porter | Snowball | Lemma)

We take all distinct words (after stopword removal) and compare the three methods.
The last column shows **DIFFERENT** where Porter and Snowball disagree.

In [13]:
all_words = []
for tokens in df["no_stopwords"]:
    for word in tokens:
        if word not in all_words:
            all_words.append(word)

rows = []
for word in all_words:
    p = porter.stem(word)
    s = snowball.stem(word)
    l = lemmatize_word(word)
    if p == s:
        check = "same"
    else:
        check = "DIFFERENT"
    rows.append([word, p, s, l, check])

compare = pd.DataFrame(rows, columns=["Original Word", "Porter Stem", "Snowball Stem", "Lemma", "Porter vs Snowball"])
compare

,Original Word,Porter Stem,Snowball Stem,Lemma,Porter vs Snowball
0,balanced,balanc,balanc,balance,same
1,diet,diet,diet,diet,same
2,helps,help,help,help,same
3,maintain,maintain,maintain,maintain,same
4,healthy,healthi,healthi,healthy,same
5,body,bodi,bodi,body,same
6,weight,weight,weight,weight,same
7,regular,regular,regular,regular,same
8,workout,workout,workout,workout,same
9,sessions,session,session,session,same


Words where Porter and Snowball disagree:

In [14]:
compare[compare["Porter vs Snowball"] == "DIFFERENT"]

,Original Word,Porter Stem,Snowball Stem,Lemma,Porter vs Snowball
35,hrs,hr,hrs,hr,DIFFERENT
50,quickly,quickli,quick,quickly,DIFFERENT


## Step 9: Vocabulary size at each stage

In [15]:
def vocab_size(column):
    words = set()
    for tokens in df[column]:
        for word in tokens:
            words.add(word)
    return len(words)

print("Raw tokens (after cleaning + tokenizing):", vocab_size("word_tokens"))
print("After stopword removal                  :", vocab_size("no_stopwords"))
print("After Porter stemming                   :", vocab_size("porter"))
print("After Snowball stemming                 :", vocab_size("snowball"))
print("After lemmatization                     :", vocab_size("lemma"))

Raw tokens (after cleaning + tokenizing): 77
After stopword removal                  : 63
After Porter stemming                   : 62
After Snowball stemming                 : 62
After lemmatization                     : 62


## Step 10: Interpretation

Stemming and lemmatization both reduced the vocabulary a little (63 → 62 words), because our dataset is small and
most words appear in only one form. The big difference is in the quality of the output: the stemmers produced
broken words like *balanc*, *healthi*, *bodi*, *daili*, *quickli* and *schedul*, while the lemmatizer gave real words
like *balance*, *healthy*, *body*, *daily* and *schedule*. Porter and Snowball gave almost the same results, Snowball
was only a little smarter (*quickly → quick*). For a search engine I would prefer **stemming** because it is fast and
it matches many forms of a word together (*improve*, *improves*, *improving* → *improv*). For a classification task or
anything where people read the output, I would prefer **lemmatization** because the words stay meaningful and
easy to understand.

## Questions for Interpretation

**1. Why can stemming produce word forms that are not valid dictionary words, while lemmatization always returns a valid word?**

A stemmer just follows fixed rules that cut off endings like *-ing*, *-es*, *-ly*, *-ity*. It does not check a
dictionary, so the result can be a word that does not exist, for example *daily → daili*, *density → densiti*,
*balanced → balanc*. A lemmatizer looks the word up in the WordNet dictionary and returns its base form (lemma),
so the output is always a real word, for example *improves → improve*, *calories → calorie*.

**2. What is the practical difference between using RegexpTokenizer and NLTK's default word_tokenize, and when would you prefer one over the other?**

`word_tokenize` is a smart tokenizer: it splits punctuation into separate tokens and handles things like
contractions (*don't → do, n't*). `RegexpTokenizer` simply keeps whatever matches the pattern we give it, so with
`[a-zA-Z]+` it returns only words and drops numbers and symbols. I would use `word_tokenize` when punctuation is
useful (for example sentence structure or sentiment like "!!"), and `RegexpTokenizer` when I only want clean words
quickly and want full control over what counts as a token.

**3. Find one word from your dataset where the Porter and Snowball stemmers give different results. Explain why the two algorithms might disagree.**

In our dataset two words give different results:
- **quickly**: Porter → *quickli*, Snowball → *quick*
- **hrs**: Porter → *hr*, Snowball → *hrs*

Snowball (also called "Porter2") is an improved version of the Porter stemmer. It has extra and updated rules,
for example a rule that removes the *-ly* ending (so *quickly* becomes *quick*), while Porter only changes the final
*y* to *i*. Snowball also does not remove the final *s* from very short words like *hrs*, but Porter does.
Because the two algorithms use slightly different rule lists, they sometimes give different stems.